### DISCUSSION TITLE
Support Service Account Key-based Authentication for Google Cloud Storage (GCS)

### DISCUSSION THREAD

**Discussion by pravallika-0506** [duckdb/duckdb#15381] [Ideas]
Currently, DuckDB supports **HMAC key-based authentication** for Google Cloud Storage (GCS). 

Link for reference: [Google Cloud Storage Import](https://duckdb.org/docs/guides/network_cloud_storage/gcs_import.html)

However, many users, including enterprise applications, rely on **Service Account Key-based authentication** for secure and best-practice access to GCS.

I would like to know if there is a way to use Service Account JSON keys in DuckDB for GCS authentication.

**Example Use Case:**

In tools like Spark, users can authenticate with service account details either by specifying the configurations directly or by providing the service account JSON key file path.

Here are two working examples using PySpark:

### Approach 1: Passing Service Account Values as Configurations

```python
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .appName("ReadFromGCS") \
    .config("spark.jars.packages", "org.apache.spark:spark-hadoop-cloud_2.12:3.4.3") \
    .config("spark.hadoop.fs.gs.impl", "com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystem") \
    .config("spark.hadoop.fs.AbstractFileSystem.gs.impl", "com.google.cloud.hadoop.fs.gcs.GoogleHadoopFS") \
    .config("spark.hadoop.google.cloud.auth.service.account.enable", "true") \
    .config("spark.hadoop.fs.gs.auth.service.account.email", "<SERVICE_ACCOUNT_EMAIL>") \
    .config("spark.hadoop.fs.gs.project.id", "<PROJECT_ID>") \
    .config("spark.hadoop.fs.gs.auth.service.account.private.key", "<PRIVATE_KEY>") \
    .config("spark.hadoop.fs.gs.auth.service.account.private.key.id", "<PRIVATE_KEY_ID>") \
    .getOrCreate()

df = spark.read.parquet("gs://<bucket-name>/<path_to_file>")
df.show()
```

### Approach 2: Using a Service Account JSON Key File

```python
import os
from pyspark.sql import SparkSession

# Path to your service account JSON key file
service_account_key_path = "/path/to/your/service-account-key.json"

spark = SparkSession.builder \
    .appName("ReadFromGCS") \
    .config("spark.jars.packages", "org.apache.spark:spark-hadoop-cloud_2.12:3.4.3") \
    .config("spark.hadoop.fs.gs.impl", "com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystem") \
    .config("spark.hadoop.fs.AbstractFileSystem.gs.impl", "com.google.cloud.hadoop.fs.gcs.GoogleHadoopFS") \
    .config("spark.hadoop.google.cloud.auth.service.account.enable", "true") \
    .config("spark.hadoop.google.cloud.auth.service.account.json.keyfile", service_account_key_path) \
    .getOrCreate()

df = spark.read.parquet("gs://<bucket-name>/<path_to_file>")
df.show()
```

If Service Account Key-based authentication is not currently supported in DuckDB, could this feature be considered for future implementation.

Thank you for your time and guidance!

**Comment by andrewhamon** [comment]
Big +1 to this. HMAC keys for GCS suck in many ways:

- less secure (long lived credential rather than an ephemeral SSO login)
- less convenient.
  - HMAC keys can only be created for a service account
  - strict limits: only 10 HMACs can be created per service account. This makes automated issuance/deletion infeasible in larger orgs
  - mature security orgs often tightly restrict who can create HMAC keys. For example, i can't create an HAMC key even if I wanted to, even though I have admin permissions on quite a few gcs buckets. I am locked out of using duckdb with gcs in my org.

All in all, HMAC keys are a very bad way to use GCS and depending on them prevents a lot of casual/exploratory use for people who work in gcloud shops.

**Comment by acornet-plume** [comment]
+1 to this as well :)

**Comment by thalassemia** [comment]
I'd also appreciate this work, because I prefer using native extensions where possible. However, I wanted to point out that for people who, like me, do not want to use HMAC keys, it is possible to use [DuckDB with `gcsfs`](https://duckdb.org/docs/stable/guides/python/filesystems), which automatically uses service account credentials. Contrary to the warning on that page, in both [this issue](https://github.com/duckdb/duckdb/issues/15140) and in my own testing, `gcsfs` is equivalent or superior in performance to `httpfs` with HMAC keys for GCS anyways.

**Comment by dentiny** [comment]
Actually there's a GCS community extension: https://github.com/northpolesec/duckdb-gcs, which seems to support SA-based auth.
https://github.com/northpolesec/duckdb-gcs?tab=readme-ov-file#authentication

**Reply by thalassemia** [reply]
Interesting! I took a look, and there is an [open issue](https://github.com/northpolesec/duckdb-gcs/issues/3) claiming the performance of that extension is 2x slower than `httpfs` using HMAC keys. Still early days though so looking forward to seeing where that extension goes.
